torch/csrc/StorageSharing.cpp

```diff
@@ -456,99 +446,32 @@ static PyObject* THPStorage_newSharedCuda(PyObject* _unused, PyObject* args) {
-  void* devPtr = basePtr.get();
-  devPtr = (char*)devPtr + storage_offset_bytes;
-
+  std::string s_handle = THPStorage_bytesAsHandleString(_handle);
   std::string ref_counter_handle = PyBytes_AS_STRING(_ref_counter);
   ptrdiff_t ref_counter_offset =
       (ptrdiff_t)THPUtils_unpackLong(_ref_counter_offset);
-
-  struct IpcDeleterContext {
-    std::string ref_counter_handle;
-    ptrdiff_t ref_counter_offset{};
-    c10::DeviceIndex device{-1};
-    torch::CudaIPCReceivedData received_data;
-  };
-
-  auto ctx = std::make_unique<IpcDeleterContext>();
-  ctx->ref_counter_handle = std::move(ref_counter_handle);
-  ctx->ref_counter_offset = ref_counter_offset;
-  ctx->device = device;
-  ctx->received_data.shared_ptr_ = std::move(basePtr);
-
-  auto cur_device = at::cuda::current_device();
-  c10::DataPtr data_ptr(
-      devPtr,
-      ctx.release(),
-      +[](void* ctx_) {
-        std::unique_ptr<IpcDeleterContext> ctx(
-            static_cast<IpcDeleterContext*>(ctx_));
-        ctx->received_data.shared_ptr_.reset();
-
-        // Sync default stream to make sure all operations related to the
-        // storage is finished (otherwise another process may reuse memory and
-        // corrupt data)
-
-        // Ideally all shared memory reference counting could be replaced by
-        // sending untriggered CUDA event from the producer to consumer and
-        // using this event as the criteria of memory release. However, CUDA
-        // (atm 10.1) does not support the creation of untriggered events and
-        // performance impact of having thousands of shared events is unknown.
-
-        // TODO: Instead of cudaStreamSynchronize it is possible to add Stream
-        // Callback and release counter inside of it (need to check performance
-        // impact)
-        at::cuda::stream_synchronize(
-            c10::cuda::getCurrentCUDAStream(ctx->device));
-
-        // We don't want to break existing code, so resource deletion is best
-        // effort basis. Exception expected if producer process terminated
-        // before consumer released data.
-        int flags =
-            at::ALLOCATOR_MAPPED_SHAREDMEM | at::ALLOCATOR_MAPPED_NOCREATE;
-        try {
-          auto sptr = at::RefcountedMapAllocator::makeDataPtr(
-              ctx->ref_counter_handle.c_str(),
-              flags,
-              sizeof(int64_t) * torch::CUDA_IPC_REF_COUNTER_FILE_SIZE,
-              nullptr);
-          *(static_cast<int64_t*>(sptr.get()) + ctx->ref_counter_offset) -= 1;
-        } catch (c10::Error& err) {
-          // Already warned inside of producer process
-        }
-      },
-      at::Device(at::DeviceType::CUDA, cur_device));
-
+  c10::DataPtr data_ptr;
+
+  at::globalContext().getAcceleratorHooksInterface().StorageNewSharedDevice(
+      device_index,
+      event_sync_required,
+      s_ipc_event_handle,
+      s_handle,
+      ref_counter_handle,
+      ref_counter_offset,
+      storage_offset_bytes,
+      data_ptr);
```

Comment:
nit: It's nice to pass mutable arguments by pointer, to make it more obvious what the in-out args are